### DISCUSSION TITLE
help needed: manual nvim-dap configuration for javascript works on windows but not on linux

### DISCUSSION THREAD

**Discussion by battlelitany** [mfussenegger/nvim-dap#1203] [General]
### Debug adapter definition and debug configuration

Hi, I am painfully new to coding/nvim in general, so please forgive me if this is a dumb question. I followed the instructions from here: [Javascript Adapter Installation](https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#javascript)

I got it all set up and working correctly on Windows 11. My configuration file is [here at pastebin](https://pastebin.com/eUeBqMWx). I am using Pop OS for Linux. I downloaded the js-debug-dap from the [vscode-js-debug](https://github.com/microsoft/vscode-js-debug) release and updated the pathway to dapDebugServer.js.

For whatever reason though, it gives me an error saying it can't connect to localhost. I'm sure it must just be something silly I forgot to do, but I can't figure out how to solve it. I tried looking up help and guides, but most of them show you how to set it up with [nvim-dap-vscode-js](https://github.com/mxsdev/nvim-dap-vscode-js). I couldn't get that to work though, and my current manual setup already works on Windows, so I'd prefer to stick with it. Any help would be much appreciated!

### Steps to Reproduce

1. Enter js file
2. Set breakpoint
3. Attempt to run

### Expected Result

It shows the nvim-dap-ui on Windows, but will not run on Linux.

### Actual Result

node exited with code 1
Couldn't connect to localhost:${port}: ECONNREFUSED

**Comment by mfussenegger** [comment]
Looks like they changed their default binding logic, and they now only bind to ipv6 `::1` instead of ipv4 `127.0.0.1` .

It should work if you change the configuration to:

```lua
dap.adapters["pwa-node"] = {
  type = "server",
  host = "::1",
  port = "${port}",
  executable = {
    command = "node",
    args = {"/path/to/js-debug/src/dapDebugServer.js", "${port}"}
  }
}
```

A good way to debug issues like this is to run the command in the terminal. E.g.:

```
node src/dapDebugServer.js 1234
```
Gave the following output:

```
Debug server listening at ::1:1234
```

**Comment by battlelitany** [comment]
> Looks like they changed their default binding logic, and they now only bind to ipv6 `::1` instead of ipv4 `127.0.0.1` .
> 
> It should work if you change the configuration to:
> 
> ```lua
> dap.adapters["pwa-node"] = {
>   type = "server",
>   host = "::1",
>   port = "${port}",
>   executable = {
>     command = "node",
>     args = {"/path/to/js-debug/src/dapDebugServer.js", "${port}"}
>   }
> }
> ```
> 
> A good way to debug issues like this is to run the command in the terminal. E.g.:
> 
> ```
> node src/dapDebugServer.js 1234
> ```
> 
> Gave the following output:
> 
> ```
> Debug server listening at ::1:1234
> ```

I tried this, but unfortunately, it still doesn't work for me. It gives me the same error: Couldn't connect to ::1:${port}: ECONNREFUSED.

Here's my edited configuration:
```
require("dap").adapters["pwa-node"] = {
   type = "server",
   host = "::1",
   port = "${port}",
   executable = {
   command = "node",
   args = { "/home/user/js-debug/src/dapDebugServer.js", "${port}" },
   },
}
```

If I run the terminal command, it gives me this output:
```
node js-debug/src/dapDebugServer.js 1234
Error: listen EADDRINUSE: address already in use ::1:1234
    at Server.setupListenHandle [as _listen2] (node:net:1898:16)
    at listenInCluster (node:net:1946:12)
    at GetAddrInfoReqWrap.doListen [as callback] (node:net:2116:7)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (node:dns:109:8) {
  code: 'EADDRINUSE',
  errno: -98,
  syscall: 'listen',
  address: '::1',
  port: 1234
}
```

EDIT: Also, I tried troubleshooting the error, but when I run `lsof -i ::1`, I get this readout:
```
lsof: unknown service :1 in: -i ::1
lsof 4.93.2
 latest revision: https://github.com/lsof-org/lsof
 latest FAQ: https://github.com/lsof-org/lsof/blob/master/00FAQ
 latest (non-formatted) man page: https://github.com/lsof-org/lsof/blob/master/Lsof.8
 usage: [-?abhKlnNoOPRtUvVX] [+|-c c] [+|-d s] [+D D] [+|-E] [+|-e s] [+|-f[gG]]
 [-F [f]] [-g [s]] [-i [i]] [+|-L [l]] [+m [m]] [+|-M] [-o [o]] [-p s]
 [+|-r [t]] [-s [p:s]] [-S [t]] [-T [t]] [-u s] [+|-w] [-x [fl]] [--] [names]

```

Thank you for such a prompt response, and apologies for troubling you!

**Comment by GhostOrder28** [comment]
@battlelitany did you manage to make it work? if not, did you tried running the adapter manually in a separate terminal?

>  home/user/js-debug/src/dapDebugServer.js

That path is a bit suspicious, in my case `dapDebugServer.js` is located in:

```
/home/<my user name>/.local/share/nvim/mason/packages/js-debug-adapter/js-debug/src/dapDebugServer.js
```
Because I installed it with Mason, how did you installed the adapter?

**Reply by battlelitany** [reply]
If this is what you mean  by running the adapter in a separate terminal, then the output for that is below!
```
node js-debug/src/dapDebugServer.js
Debug server listening at ::1:8123
```

The path is specified that way because I installed the adapter manually! On the instructions page from [here for manual installation](https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#javascript), it said to download and extract to any folder.

I did so on Windows by putting it at the bottom of my C: drive and that worked perfectly! Since the instructions didn't seem to say if they were Windows/Linux specific, I assumed the same process would work on Linux. As such, I also extracted it and put it at the base of my Home folder.

Again, apologies for the trouble. I'm new to coding and Linux and a lot of it is just me trying to fumble around to teach myself things. I'd like to apologize if I missed something obvious while trying to troubleshoot this. Thank you for your help!

**Comment by GhostOrder28** [comment]
> The path is specified that way because I installed the adapter manually!

Gotcha, I asked just in case because one can sometimes make typos when writing paths.

Can you try modifying your adapter config this way?
```
require("dap").adapters["pwa-node"] = {
   type = "server",
   host = "::1",
   port = "8123
}
```
Then first run the adapter manually and then try to start a debug session in nvim.

Also can you share your config for `require("dap").configurations.<your language>`

**Reply by battlelitany** [reply]
This works! It still strangely gives a "node exited with code 1" error, but the nvim-dap does indeed launch and run properly. Thank you! I'm not sure why it still gives the error though.

Here's the config of the part I edited and my language:

```
    require("dap").adapters["pwa-node"] = {
      type = "server",
      host = "::1",
      port = "8123",
      executable = {
        command = "node",
        args = {
          "C:/home/cafe/js-debug/src/dapDebugServer.js",
          "${port}",
        },
      },
    }

    require("dap").configurations.javascript = {
      {
        type = "pwa-node",
        request = "launch",
        name = "Launch file",
        program = "${file}",
        cwd = "${workspaceFolder}",
        console = "integratedTerminal", --this keeps the terminal from creating a new window
      },
    }
```

**Reply by battlelitany** [reply]
Had the chance to test this configuration again on a different linux computer (same OS though) and it still gives the same "node exited with code 1" error but will run. I still have to run the adapter manually first in the terminal though!

I'm happy it works, but is there any way I can skip having to run the adapter manually first?
